test/test_custom_ops.py

```diff
@@ -2171,6 +2171,38 @@ def test_impl_device_invalid(self):
         with self.assertRaisesRegex(RuntimeError, "Expected one of cpu, cuda"):
             torch.library.impl("blah::blah", "somethingsomething")
 
+    def test_duplicate_fake(self):
+        lib = self.lib()
+        op_name = f"{self.test_ns}::foo"
+        torch.library.define(op_name, "(Tensor x) -> Tensor", lib=lib)
+        op = self.ns().foo.default
+
+        def foo_impl1(x):
+            return x * 1
+
+        lib.impl("foo", foo_impl1, "CPU")
+        torch.library.register_fake(op_name, foo_impl1, lib=lib)
+
+        self.assertEqual(op(torch.ones(3)), torch.ones(3))
+        with torch._subclasses.FakeTensorMode():
+            self.assertEqual(op(torch.ones(3)).shape, [3])
+
+        def foo_impl2(x):
+            return torch.cat([x, x])
+
+        with self.assertRaisesRegex(RuntimeError, "already a kernel registered"):
+            lib.impl("foo", foo_impl2, "CPU")
+
+        with self.assertRaisesRegex(RuntimeError, "already has an fake impl"):
+            torch.library.register_fake(op_name, foo_impl2, lib=lib)
+
+        torch.library.register_fake(op_name, foo_impl2, lib=lib, _allow_override=True)
+        with torch._subclasses.FakeTensorMode():
+            self.assertEqual(op(torch.ones(3)).shape, [6])
+
+        lib.impl(op_name, foo_impl2, "CPU", _allow_override=True)
+        self.assertEqual(op(torch.ones(3)), torch.ones(6))
+
```

Comment:
Can we get tests for:
1) can you check that duplicate register_fake also overrides the meta kernel? e.g. use the above tests but send in a device="meta" tensor instead of a FakeTensor.
2) what happens when you torch.library.impl("foo", fn, "Meta") twice with different fns? 
3) what happens when you do the following:
```
lib1 = Library("mylib", "FRAGMENT")
lib1.impl("foo", fn1, "Meta")
with scoped_library("mylib", "FRAGMENT"):
    lib2.impl("foo", fn2, "Meta")
foo(x)  # should be using fn1
```
4) same as (3), but for register_fake
5) register a kernel to compositeexplicitautograd and then use register_fake to override it
6) register a kernel to meta and then use register_fake to override it

Reply:
Added the tests above except for the scoped_library tests. I'm a little confused if that should be the behavior for scoped_library. This means that we would have to record all the previous impls and then re-register them. For the profile context manager in my following diff it's easier because we just need to keep track of the fake_impls.

Reply:
That is indeed the behavior for scoped_library today where we do not block duplicate registrations:
![image](https://github.com/user-attachments/assets/bf786b1c-faa6-4710-963a-054d5fc9a38f)